File: projects/lua/assets/shooter.lua
local PLAYER_SPEED = 330
local BULLET_SPEED = 570
local ENEMY_SPEED = 145
local FIRE_DELAY = 0.18
local SPAWN_DELAY = 0.72

local player = { x = 0, y = -300 }
local bullets = {}
local enemies = {}
local score = 0
local lives = 3
local next_id = 1
local fire_timer = 0
local spawn_timer = 0
local seed = 73129
local game_over = false
local was_firing = false

local function spawn_entity(kind, id, x, y)
    ecs_spawn_entity(id)
    ecs_insert_sprite(id, kind)
    ecs_set_transform(id, x, y)
end

local function random01()
    seed = (seed * 48271) % 2147483647
    return seed / 2147483647
end

local function reset_game()
    ecs_clear_world()
    player = { x = 0, y = -300 }
    bullets = {}
    enemies = {}
    score = 0
    lives = 3
    next_id = 1
    fire_timer = 0
    spawn_timer = 0.35
    seed = 73129
    game_over = false
    spawn_entity("player", "player", player.x, player.y)
    ecs_set_game_state(score, lives, "ARROWS/WASD + HOLD SPACE")
end

local function spawn_enemy()
    local enemy = {
        id = "enemy_" .. next_id,
        x = -250 + random01() * 500,
        y = 350,
        alive = true,
    }
    next_id = next_id + 1
    table.insert(enemies, enemy)
    spawn_entity("enemy", enemy.id, enemy.x, enemy.y)
end

local function shoot()
    local bullet = {
        id = "bullet_" .. next_id,
        x = player.x,
        y = player.y + 50,
        alive = true,
    }
    next_id = next_id + 1
    table.insert(bullets, bullet)
    spawn_entity("bullet", bullet.id, bullet.x, bullet.y)
end

local function hit(a, b, half_width, half_height)
    return math.abs(a.x - b.x) < half_width and math.abs(a.y - b.y) < half_height
end

function on_script_loaded()
    reset_game()
end

function on_script_reloaded()
    reset_game()
end

function on_update(dt, input_x, input_y, firing)
    if game_over then
        if firing and not was_firing then reset_game() end
        was_firing = firing
        return
    end

    player.x = math.max(-260, math.min(260, player.x + input_x * PLAYER_SPEED * dt))
    player.y = math.max(-335, math.min(300, player.y + input_y * PLAYER_SPEED * dt))
    ecs_set_transform("player", player.x, player.y)

    fire_timer = fire_timer - dt
    if firing and fire_timer <= 0 then
        shoot()
        fire_timer = FIRE_DELAY
    end

    spawn_timer = spawn_timer - dt
    if spawn_timer <= 0 then
        spawn_enemy()
        spawn_timer = SPAWN_DELAY
    end

    for _, bullet in ipairs(bullets) do
        bullet.y = bullet.y + BULLET_SPEED * dt
        if bullet.y > 420 then
            bullet.alive = false
            ecs_despawn_entity(bullet.id)
        else
            ecs_set_transform(bullet.id, bullet.x, bullet.y)
        end
    end

    for _, enemy in ipairs(enemies) do
        enemy.y = enemy.y - ENEMY_SPEED * dt
        if enemy.y < -420 then
            enemy.alive = false
            lives = lives - 1
            ecs_despawn_entity(enemy.id)
        else
            ecs_set_transform(enemy.id, enemy.x, enemy.y)
        end
    end

    for _, bullet in ipairs(bullets) do
        if bullet.alive then
            for _, enemy in ipairs(enemies) do
                if enemy.alive and hit(bullet, enemy, 36, 42) then
                    bullet.alive = false
                    enemy.alive = false
                    score = score + 100
                    ecs_despawn_entity(bullet.id)
                    ecs_despawn_entity(enemy.id)
                    break
                end
            end
        end
    end

    for _, enemy in ipairs(enemies) do
        if enemy.alive and hit(player, enemy, 55, 65) then
            enemy.alive = false
            lives = lives - 1
            ecs_despawn_entity(enemy.id)
        end
    end

    for i = #bullets, 1, -1 do
        if not bullets[i].alive then table.remove(bullets, i) end
    end
    for i = #enemies, 1, -1 do
        if not enemies[i].alive then table.remove(enemies, i) end
    end

    if lives <= 0 then
        lives = 0
        game_over = true
        ecs_set_game_state(score, lives, "GAME OVER - TAP SPACE TO RESTART")
    else
        ecs_set_game_state(score, lives, "")
    end
    was_firing = firing
end

File: projects/ts/assets/shooter.js
"use strict";
(() => {
  // projects/ts/src/ecs.ts
  function clearWorld() {
    ecs_world_clear();
  }
  function spawnEntity(id) {
    ecs_entity_spawn(id);
  }
  function despawnEntity(id) {
    return ecs_entity_despawn(id);
  }
  function insertComponent(id, name, value) {
    return ecs_component_insert(id, name, value);
  }
  function setResource(name, value) {
    ecs_resource_set(name, value);
  }

  // projects/ts/src/shooter.ts
  var PLAYER_SPEED = 330;
  var BULLET_SPEED = 570;
  var ENEMY_SPEED = 145;
  var FIRE_DELAY = 0.18;
  var SPAWN_DELAY = 0.72;
  var DAMAGE_DELAY = 1;
  function createWorld() {
    return {
      entities: /* @__PURE__ */ new Set(),
      transforms: /* @__PURE__ */ new Map(),
      velocities: /* @__PURE__ */ new Map(),
      colliders: /* @__PURE__ */ new Map(),
      sprites: /* @__PURE__ */ new Map(),
      players: /* @__PURE__ */ new Set(),
      bullets: /* @__PURE__ */ new Set(),
      enemies: /* @__PURE__ */ new Set(),
      pendingDespawn: /* @__PURE__ */ new Set()
    };
  }
  function createResources() {
    return {
      score: 0,
      lives: 3,
      nextId: 1,
      fireTimer: 0,
      spawnTimer: 0.35,
      damageTimer: 0,
      seed: 73129,
      gameOver: false,
      restartWasPressed: false,
      started: false
    };
  }
  var world = createWorld();
  var resources = createResources();
  function spawnEntity2(id, bundle) {
    world.entities.add(id);
    world.transforms.set(id, bundle.transform);
    world.colliders.set(id, bundle.collider);
    world.sprites.set(id, bundle.sprite);
    if (bundle.velocity) world.velocities.set(id, bundle.velocity);
    if (bundle.role === "player") world.players.add(id);
    if (bundle.role === "bullet") world.bullets.add(id);
    if (bundle.role === "enemy") world.enemies.add(id);
    spawnEntity(id);
    insertComponent(id, "sprite", { kind: bundle.sprite });
    insertComponent(id, "transform", {
      x: bundle.transform.x,
      y: bundle.transform.y
    });
  }
  function queueDespawn(id) {
    if (world.entities.has(id)) world.pendingDespawn.add(id);
  }
  function isActive(id) {
    return world.entities.has(id) && !world.pendingDespawn.has(id);
  }
  function flushEntityCommands() {
    for (const id of world.pendingDespawn) {
      world.entities.delete(id);
      world.transforms.delete(id);
      world.velocities.delete(id);
      world.colliders.delete(id);
      world.sprites.delete(id);
      world.players.delete(id);
      world.bullets.delete(id);
      world.enemies.delete(id);
      despawnEntity(id);
    }
    world.pendingDespawn.clear();
  }
  function random01() {
    resources.seed = resources.seed * 48271 % 2147483647;
    return resources.seed / 2147483647;
  }
  function spawnPlayer() {
    spawnEntity2("player", {
      role: "player",
      sprite: "player",
      transform: { x: 0, y: -300 },
      collider: { x: 25, y: 35 }
    });
  }
  function spawnEnemy() {
    spawnEntity2(`enemy_${resources.nextId++}`, {
      role: "enemy",
      sprite: "enemy",
      transform: { x: -250 + random01() * 500, y: 350 },
      velocity: { x: 0, y: -ENEMY_SPEED },
      collider: { x: 30, y: 30 }
    });
  }
  function spawnBullet(playerTransform) {
    spawnEntity2(`bullet_${resources.nextId++}`, {
      role: "bullet",
      sprite: "bullet",
      transform: { x: playerTransform.x, y: playerTransform.y + 50 },
      velocity: { x: 0, y: BULLET_SPEED },
      collider: { x: 6, y: 12 }
    });
  }
  function playerMovementSystem(frame) {
    for (const id of world.players) {
      const transform = world.transforms.get(id);
      if (!transform || !isActive(id)) continue;
      transform.x = Math.max(-260, Math.min(260, transform.x + frame.inputX * PLAYER_SPEED * frame.dt));
      transform.y = Math.max(-335, Math.min(300, transform.y + frame.inputY * PLAYER_SPEED * frame.dt));
    }
  }
  function weaponSystem(frame) {
    resources.fireTimer -= frame.dt;
    if (resources.fireTimer > 0) return;
    for (const id of world.players) {
      const transform = world.transforms.get(id);
      if (transform && isActive(id)) spawnBullet(transform);
    }
    resources.fireTimer = FIRE_DELAY;
  }
  function enemySpawnSystem(frame) {
    resources.spawnTimer -= frame.dt;
    if (resources.spawnTimer <= 0) {
      spawnEnemy();
      resources.spawnTimer = SPAWN_DELAY;
    }
  }
  function movementSystem(frame) {
    for (const [id, velocity] of world.velocities) {
      const transform = world.transforms.get(id);
      if (!transform || !isActive(id)) continue;
      transform.x += velocity.x * frame.dt;
      transform.y += velocity.y * frame.dt;
    }
  }
  function boundsSystem() {
    for (const id of world.bullets) {
      const transform = world.transforms.get(id);
      if (transform && transform.y > 420) queueDespawn(id);
    }
    for (const id of world.enemies) {
      const transform = world.transforms.get(id);
      if (transform && transform.y < -420) queueDespawn(id);
    }
  }
  function entitiesOverlap(left, right) {
    const leftTransform = world.transforms.get(left);
    const rightTransform = world.transforms.get(right);
    const leftCollider = world.colliders.get(left);
    const rightCollider = world.colliders.get(right);
    if (!leftTransform || !rightTransform || !leftCollider || !rightCollider) return false;
    return Math.abs(leftTransform.x - rightTransform.x) < leftCollider.x + rightCollider.x && Math.abs(leftTransform.y - rightTransform.y) < leftCollider.y + rightCollider.y;
  }
  function collisionSystem(frame) {
    for (const bullet of world.bullets) {
      if (!isActive(bullet)) continue;
      for (const enemy of world.enemies) {
        if (isActive(enemy) && entitiesOverlap(bullet, enemy)) {
          queueDespawn(bullet);
          queueDespawn(enemy);
          resources.score += 100;
          break;
        }
      }
    }
    resources.damageTimer = Math.max(0, resources.damageTimer - frame.dt);
    if (resources.damageTimer > 0) return;
    for (const player of world.players) {
      if (!isActive(player)) continue;
      for (const enemy of world.enemies) {
        if (isActive(enemy) && entitiesOverlap(player, enemy)) {
          queueDespawn(enemy);
          resources.lives -= 1;
          resources.damageTimer = DAMAGE_DELAY;
          return;
        }
      }
    }
  }
  function renderSyncSystem() {
    for (const [id, transform] of world.transforms) {
      if (isActive(id)) insertComponent(id, "transform", transform);
    }
  }
  function gameStateSystem() {
    if (resources.lives <= 0) {
      resources.lives = 0;
      resources.gameOver = true;
      setResource("game_state", {
        score: resources.score,
        lives: resources.lives,
        message: "GAME OVER - TAP SPACE TO RESTART"
      });
    } else {
      setResource("game_state", {
        score: resources.score,
        lives: resources.lives,
        message: ""
      });
    }
  }
  var updateSchedule = [
    playerMovementSystem,
    weaponSystem,
    enemySpawnSystem,
    movementSystem,
    boundsSystem,
    collisionSystem
  ];
  function resetGame() {
    clearWorld();
    world = createWorld();
    resources = createResources();
    spawnPlayer();
    setResource("game_state", {
      score: resources.score,
      lives: resources.lives,
      message: "ARROWS/WASD - AUTO FIRE"
    });
  }
  var callbacks = globalThis;
  callbacks.on_script_loaded = function() {
    resetGame();
  };
  callbacks.on_script_reloaded = function() {
    resetGame();
  };
  callbacks.on_update = function(dt, inputX, inputY, restartPressed) {
    if (!resources.started) {
      if (restartPressed && !resources.restartWasPressed) {
        resources.started = true;
        setResource("game_state", {
          score: resources.score,
          lives: resources.lives,
          message: "ARROWS/WASD - AUTO FIRE"
        });
      } else {
        resources.restartWasPressed = restartPressed;
        setResource("game_state", {
          score: resources.score,
          lives: resources.lives,
          message: "PRESS SPACE TO START"
        });
        return;
      }
    }
    if (resources.gameOver) {
      if (restartPressed && !resources.restartWasPressed) resetGame();
      resources.restartWasPressed = restartPressed;
      return;
    }
    const frame = { dt, inputX, inputY };
    for (const system of updateSchedule) system(frame);
    flushEntityCommands();
    gameStateSystem();
    renderSyncSystem();
    resources.restartWasPressed = restartPressed;
  };
})();
